module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -215,6 +217,139 @@ func TestInvalidateEntity(t *testing.T) {
+
+	// an array of random owner Ids.
+	ownerIds := make([]uint64, entityCount, entityCount)
+	// generate ownerId to index in the entities arrray.
+	for i := 0; i < int(entityCount); i++ {
+		ownerIds[i] = rand.Uint64()
+	}
+	// this map maintains entities currently stored in the pool.
+	addedEntities := make(map[flow.Identifier]int)
+	addedEntitiesInPool := make(map[flow.Identifier]EIndex)
+
+	for i := 0; i < numberOfOperations; i++ {
+		// choose between Add and Remove with a probability of probabilityOfAdding and 1-probabilityOfAdding respectively.
+		if rand.Float32() < probabilityOfAdding || len(addedEntities) == 0 {
+
+			// find an entity to add
+			entityToAddIndex := -1
+			found := false
+			for retryTime := 0; retryTime < 100; retryTime++ {
+				entityToAddIndexTmp := rand.Intn(int(entityCount))
+				_, found = addedEntities[entities[entityToAddIndexTmp].ID()]
+				if !found {
+					entityToAddIndex = entityToAddIndexTmp
+					break
+				}
+			}
+			if !found {
+				indexInThePool, _, ejectedEntity := pool.Add(entities[entityToAddIndex].ID(), entities[entityToAddIndex], uint64(ownerIds[entityToAddIndex]))
+				require.True(t, indexInThePool != InvalidIndex || (ejectionMode == NoEjection && len(addedEntities) == int(limit)))
+				require.True(t, ejectionMode != NoEjection || ejectedEntity == nil)
+				if indexInThePool != InvalidIndex {
+					addedEntities[entities[entityToAddIndex].ID()] = entityToAddIndex
+					addedEntitiesInPool[entities[entityToAddIndex].ID()] = indexInThePool
+					actualFlowId, actualEntity, actualOwnerId := pool.Get(indexInThePool)
+					require.Equal(t, entities[entityToAddIndex].ID(), actualFlowId)
+					require.Equal(t, entities[entityToAddIndex], actualEntity)
+					require.Equal(t, ownerIds[entityToAddIndex], actualOwnerId)
+				}
+				if ejectedEntity != nil {
+					require.Contains(t, addedEntities, ejectedEntity.ID(), "pool ejected an entity that was not added before")
+					delete(addedEntities, ejectedEntity.ID())
+					delete(addedEntitiesInPool, ejectedEntity.ID())
+				}
+			}
+		} else {
+			// randomly select an index of an entity to remove.
+			entityToRemove := rand.Intn(len(addedEntities))
+			i := 0
+			var indexInPoolToRemove EIndex = 0
+			var indexInEntitiesArray int = 0
+			for k, v := range addedEntities {
+				if i == entityToRemove {
+					indexInPoolToRemove = addedEntitiesInPool[k]
+					indexInEntitiesArray = v
+					break
+				}
+				i++
+			}
+			// remove the selected entity from the pool.
+			removedEntity := pool.Remove(indexInPoolToRemove)
+			require.Equal(t, entities[indexInEntitiesArray].ID(), removedEntity.ID(), "removed wrong entity")
+			delete(addedEntities, entities[indexInEntitiesArray].ID())
+			delete(addedEntitiesInPool, entities[indexInEntitiesArray].ID())
+			actualFlowId, actualEntity, _ := pool.Get(indexInPoolToRemove)
+			require.Equal(t, flow.ZeroID, actualFlowId)
+			require.Equal(t, nil, actualEntity)
+		}
+	}
```

Comment:
It's like to say "please code better" :). No, but if seriously, I'll take a look at the diff and will try to understand on what I can improve.

I've already stated that this condition
```require.Falsef(t, found, "could not find an entity to add after %d retries", retryLimit)``` might constantly fail. And for that reason I have not included it in my diff.
Please try 
```
			limit:               1000,
			entityCount:         500,
```
you will see that the test never passes. Please, let me know if this is intended. If you'd like to keep 
```require.Falsef(t, found, "could not find an entity to add after %d retries", retryLimit)```
I think we need at least to comment in the test description that by design it cannot work for certain limit and entityCount combination. Basically, once all the entities are inserted it's not possible to find an "uninsured" entity.

Reply:
@yhassanzadeh13 Please check a couple of minor comments that I had to your proposed refactoring. And a concern about failing test comment above.

I have a version with your changes included ready on my local drive. Once that I have your answers I'll proceed with submitting a new patch.

Reply:
@kozlovb thanks for the reply. It is acceptable to limit the scope of this test to cases where `limit` <= `entityCount`. In order to ensure that this condition is treated as an invariant, I suggest adding the following check at the beginning of the test: 

```go
require.GreaterOrEqual(t, entityCount, limit, "entityCount must be greater or equal to limit to test add/remove operations")
```

Reply:
Unfortunately it's not as clear cut. The problem is that when limit and entityCount are getting close it can also break but it becomes probabilistic. Imagine the limit is 1000 and entityCount is 1001. In this case out of 100 tries the probability is only 10% to find that the 1 non inserted entity is found. Let's put that the difference has to be at least factor of 2. I'll update the PR now and we from there. Sounds good ?

Reply:
@yhassanzadeh13 I pushed the latest diff without applying my 3 minor suggestions from https://github.com/onflow/flow-go/pull/4470#discussion_r1263680127
Please, check them out.

Reply:
@yhassanzadeh13 I think now everything is in. The change ended up being quite big. Overall, I feel confident but I think it would make sense for me to take 1-2h on Monday morning to thoroughly check everything one more time.

Reply:
@yhassanzadeh13 I went through some closed threads today. I did some minor clean-up. Looks good to me now for the review from your side. Please take another look.